farmerbot/internal/find_node.go

```diff
@@ -0,0 +1,154 @@
+			return 0, fmt.Errorf("not enough public ips available for farm %d", f.farm.ID)
+		}
+	}
+
+	var possibleNodes []node
+	for _, node := range f.nodes {
+		gpus := node.gpus
+		if nodeOptions.HasGPUs > 0 {
+			if len(nodeOptions.GPUVendors) > 0 {
+				gpus = filterGPUs(gpus, nodeOptions.GPUVendors, false)
+			}
+
+			if len(nodeOptions.GPUDevices) > 0 {
+				gpus = filterGPUs(gpus, nodeOptions.GPUDevices, true)
+			}
+
+			if len(gpus) < int(nodeOptions.HasGPUs) {
+				continue
+			}
+		}
+
+		if nodeOptions.Certified && !node.Certification.IsCertified {
+			continue
+		}
+
+		if nodeOptions.PublicConfig && !node.PublicConfig.HasValue {
+			continue
+		}
+
+		if node.hasActiveRentContract {
+			continue
+		}
+
+		if nodeOptions.Dedicated {
+			if !node.dedicated || !node.isUnused() {
+				continue
+			}
+		} else {
+			if node.dedicated && nodeOptionsCapacity != node.resources.total {
+				continue
+			}
+		}
+
+		if slices.Contains(nodeOptions.NodeExclude, uint32(node.ID)) {
+			continue
+		}
+
+		if !node.canClaimResources(nodeOptionsCapacity, f.config.Power.OverProvisionCPU) {
+			continue
+		}
+
+		possibleNodes = append(possibleNodes, node)
+	}
+
+	if len(possibleNodes) == 0 {
+		return 0, fmt.Errorf("could not find a suitable node with the given options: %+v", nodeOptions)
+	}
+
+	// Sort the nodes on power state (the ones that are ON first then waking up, off, shutting down)
+	sort.Slice(possibleNodes, func(i, j int) bool {
+		return possibleNodes[i].ID < possibleNodes[j].ID
+	})
+
+	sort.Slice(possibleNodes, func(i, j int) bool {
+		return possibleNodes[i].powerState < possibleNodes[j].powerState
+	})
+
+	nodeFound := possibleNodes[0]
+	log.Debug().Uint32("nodeID", uint32(nodeFound.ID)).Msg("Found a node")
+
+	// claim the resources until next update of the data
+	// add a timeout (after 30 minutes we update the resources)
+	nodeFound.timeoutClaimedResources = time.Now().Add(timeoutPowerStateChange)
+
+	if nodeOptions.Dedicated || nodeOptions.HasGPUs > 0 {
+		// claim all capacity
+		nodeFound.claimResources(nodeFound.resources.total)
+	} else {
+		nodeFound.claimResources(nodeOptionsCapacity)
+	}
+
+	// claim public ips until next update of the data
+	if nodeOptions.PublicIPs > 0 {
+		nodeFound.publicIPsUsed += nodeOptions.PublicIPs
+	}
+
+	// power on the node if it is down or if it is shutting down
+	if nodeFound.powerState == off || nodeFound.powerState == shuttingDown {
+		if err := f.powerOn(sub, uint32(nodeFound.ID)); err != nil {
+			return 0, fmt.Errorf("failed to power on found node %d", nodeFound.ID)
+		}
+	}
+
+	// update claimed resources
+	err := f.updateNode(nodeFound)
+	if err != nil {
+		return 0, fmt.Errorf("failed to power on found node %d", nodeFound.ID)
+	}
+
+	return uint32(nodeFound.ID), nil
+}
+
+// FilterIncludesSubStr filters a string slice according to if elements include a sub string
+func filterGPUs(gpus []zos.GPU, vendorsOrDevices []string, device bool) (filtered []zos.GPU) {
```

Comment:
this should have been splitted to two functions imo filterGPUsByVendors and filterGPUsByDevices